### PR TITLE
Pbi 10 view detail list monitor backend ferdi

### PR DESCRIPTION
**Pull request by ferdifdi**
### Describe your changes
Detail list monitor that consists retrieving of success rate and response time for 30min, 1h, 3h, 6h, 12h, and 24h

### Backlog name
PBI-10-View_API_Monitor_detail

### Acceptance criteria
- Detail API Monitor shows success rate and response time with chosen period

### Example Request
```
curl --location --request GET 'localhost:8000/monitor/9/?range=30MIN' \
--header 'Authorization: Token 94accc6232e9aef7f098b4e2a6afef19ddbae295'

{
    "id": 9,
    "name": "Test Create",
    "method": "GET",
    "url": "test/api",
    "schedule": "30MIN",
    "body_type": "RAW",
    "query_params": [
        {
            "id": 15,
            "monitor": 9,
            "key": "qp_key1",
            "value": "qp_value1"
        },
        {
            "id": 16,
            "monitor": 9,
            "key": "qp_key2",
            "value": "qp_value2"
        }
    ],
    "headers": [
        {
            "id": 8,
            "monitor": 9,
            "key": "h_key1",
            "value": "h_value1"
        }
    ],
    "body_form": [],
    "raw_body": {
        "id": 8,
        "monitor": 9,
        "body": "Test raw body"
    },
    "success_rate": [
        {
            "start_time": "2022-10-11T23:55:54.784248+07:00",
            "end_time": "2022-10-11T23:56:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-11T23:56:54.784248+07:00",
            "end_time": "2022-10-11T23:57:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-11T23:57:54.784248+07:00",
            "end_time": "2022-10-11T23:58:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-11T23:58:54.784248+07:00",
            "end_time": "2022-10-11T23:59:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-11T23:59:54.784248+07:00",
            "end_time": "2022-10-12T00:00:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:00:54.784248+07:00",
            "end_time": "2022-10-12T00:01:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:01:54.784248+07:00",
            "end_time": "2022-10-12T00:02:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:02:54.784248+07:00",
            "end_time": "2022-10-12T00:03:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:03:54.784248+07:00",
            "end_time": "2022-10-12T00:04:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:04:54.784248+07:00",
            "end_time": "2022-10-12T00:05:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:05:54.784248+07:00",
            "end_time": "2022-10-12T00:06:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:06:54.784248+07:00",
            "end_time": "2022-10-12T00:07:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:07:54.784248+07:00",
            "end_time": "2022-10-12T00:08:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:08:54.784248+07:00",
            "end_time": "2022-10-12T00:09:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:09:54.784248+07:00",
            "end_time": "2022-10-12T00:10:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:10:54.784248+07:00",
            "end_time": "2022-10-12T00:11:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:11:54.784248+07:00",
            "end_time": "2022-10-12T00:12:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:12:54.784248+07:00",
            "end_time": "2022-10-12T00:13:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:13:54.784248+07:00",
            "end_time": "2022-10-12T00:14:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:14:54.784248+07:00",
            "end_time": "2022-10-12T00:15:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:15:54.784248+07:00",
            "end_time": "2022-10-12T00:16:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:16:54.784248+07:00",
            "end_time": "2022-10-12T00:17:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:17:54.784248+07:00",
            "end_time": "2022-10-12T00:18:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:18:54.784248+07:00",
            "end_time": "2022-10-12T00:19:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:19:54.784248+07:00",
            "end_time": "2022-10-12T00:20:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:20:54.784248+07:00",
            "end_time": "2022-10-12T00:21:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:21:54.784248+07:00",
            "end_time": "2022-10-12T00:22:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:22:54.784248+07:00",
            "end_time": "2022-10-12T00:23:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:23:54.784248+07:00",
            "end_time": "2022-10-12T00:24:54.784248+07:00",
            "success": 0,
            "failed": 0
        },
        {
            "start_time": "2022-10-12T00:24:54.784248+07:00",
            "end_time": "2022-10-12T00:25:54.784248+07:00",
            "success": 0,
            "failed": 0
        }
    ],
    "response_time": [
        {
            "start_time": "2022-10-11T23:55:54.784248+07:00",
            "end_time": "2022-10-11T23:56:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-11T23:56:54.784248+07:00",
            "end_time": "2022-10-11T23:57:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-11T23:57:54.784248+07:00",
            "end_time": "2022-10-11T23:58:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-11T23:58:54.784248+07:00",
            "end_time": "2022-10-11T23:59:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-11T23:59:54.784248+07:00",
            "end_time": "2022-10-12T00:00:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:00:54.784248+07:00",
            "end_time": "2022-10-12T00:01:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:01:54.784248+07:00",
            "end_time": "2022-10-12T00:02:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:02:54.784248+07:00",
            "end_time": "2022-10-12T00:03:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:03:54.784248+07:00",
            "end_time": "2022-10-12T00:04:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:04:54.784248+07:00",
            "end_time": "2022-10-12T00:05:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:05:54.784248+07:00",
            "end_time": "2022-10-12T00:06:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:06:54.784248+07:00",
            "end_time": "2022-10-12T00:07:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:07:54.784248+07:00",
            "end_time": "2022-10-12T00:08:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:08:54.784248+07:00",
            "end_time": "2022-10-12T00:09:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:09:54.784248+07:00",
            "end_time": "2022-10-12T00:10:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:10:54.784248+07:00",
            "end_time": "2022-10-12T00:11:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:11:54.784248+07:00",
            "end_time": "2022-10-12T00:12:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:12:54.784248+07:00",
            "end_time": "2022-10-12T00:13:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:13:54.784248+07:00",
            "end_time": "2022-10-12T00:14:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:14:54.784248+07:00",
            "end_time": "2022-10-12T00:15:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:15:54.784248+07:00",
            "end_time": "2022-10-12T00:16:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:16:54.784248+07:00",
            "end_time": "2022-10-12T00:17:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:17:54.784248+07:00",
            "end_time": "2022-10-12T00:18:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:18:54.784248+07:00",
            "end_time": "2022-10-12T00:19:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:19:54.784248+07:00",
            "end_time": "2022-10-12T00:20:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:20:54.784248+07:00",
            "end_time": "2022-10-12T00:21:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:21:54.784248+07:00",
            "end_time": "2022-10-12T00:22:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:22:54.784248+07:00",
            "end_time": "2022-10-12T00:23:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:23:54.784248+07:00",
            "end_time": "2022-10-12T00:24:54.784248+07:00",
            "avg": 0
        },
        {
            "start_time": "2022-10-12T00:24:54.784248+07:00",
            "end_time": "2022-10-12T00:25:54.784248+07:00",
            "avg": 0
        }
    ]
}
```

### Checklist
- [X] I have performed a self-review of my code
- [X] Code successfully run on local
- [X] Feature / changes tested on local
- [X] No failing unit test
- [ ] Passed UAT Test
- [ ] Sonarqube tested 
- [ ] Required any changes to environment variable

### CODE REVIEW

File: apimonitor/views.py [hunk 1]
```diff
@@ -118,6 +121,72 @@ def create(self, request, *args, **kwargs):
     def get_queryset(self):
         queryset = APIMonitor.objects.filter(user=self.request.user)
         return queryset
+
+    def retrieve(self, request, pk):          
+
+        queryset = self.filter_queryset(self.get_queryset())
+
+        monitor = get_object_or_404(queryset,pk=pk)
```
Comment (andrew-susanto) [inline apimonitor/views.py:129]:
```suggestion
    def retrieve(self, request, pk=None):          
        queryset = self.filter_queryset(self.get_queryset())
        monitor = get_object_or_404(queryset,pk=pk)
```

nitip diubah pk=None ferr biar sesuai sama docsnya wkkwk
https://www.django-rest-framework.org/api-guide/viewsets/

File: apimonitor/views.py [hunk 2]
```diff
@@ -118,6 +121,72 @@ def create(self, request, *args, **kwargs):
...
+
+        response_time = []
+        success_rate = []
+
+        def retrieve_with_param(last_chosen_period_in_hours, n_bar, increment_in_minutes):
```
Comment (andrew-susanto) [inline apimonitor/views.py:134]: ini coba ditaruh di luar function aja kali (?) wkwkk kayaknya gaperlu nested bisa

**Review by HugoIr — COMMENTED** [review]
one code smells

**Review by HugoIr — APPROVED** [review]

File: apimonitor/views.py [hunk 3]
```diff
@@ -119,6 +122,71 @@ def get_queryset(self):
         queryset = APIMonitor.objects.filter(user=self.request.user)
         return queryset
     
+    def retrieve_with_param(last_chosen_period_in_hours, n_bar, increment_in_minutes, monitor, success_rate, response_time):
```
Comment (HugoIr) [inline apimonitor/views.py:125]: One code smell on this function. Bisa coba direname "last_chosen_period_in_hours" to "self" or add the missing "self" parameter?

**Review by HugoIr — COMMENTED** [review]
one code smell

**Review by andrew-susanto — APPROVED** [review]

**Review by luthfifahlevi — APPROVED** [review]

**Pull request merged**